Fix potencial leak / seg fault

ahcorde opened the pull request:
Signed-off-by: ahcorde <ahcorde@gmail.com>

This PR is a follow up of this other PR https://github.com/ros2/rviz/pull/701

ahcorde committed:
Fix potencial leak / seg fault

Signed-off-by: ahcorde <ahcorde@gmail.com>

ahcorde commented:
* Linux [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux&build=14886)](http://ci.ros2.org/job/ci_linux/14886/)
* Linux-aarch64 [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux-aarch64&build=9634)](http://ci.ros2.org/job/ci_linux-aarch64/9634/)
* macOS [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_osx&build=12553)](http://ci.ros2.org/job/ci_osx/12553/)
* Windows [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_windows&build=15044)](http://ci.ros2.org/job/ci_windows/15044/)

clalancette commented on rviz_default_plugins/src/rviz_default_plugins/displays/image/ros_image_texture.cpp:
```diff
@@ -418,20 +418,36 @@ ImageData ROSImageTexture::setFormatAndNormalizeDataIfNecessary(ImageData image_
     std::vector<uint8_t> buffer = normalize<uint16_t>(
       reinterpret_cast<const uint16_t *>(image_data.data_ptr_),
       image_data.size_);
+    std::copy(buffer.begin(), buffer.end(), bufferptr_->begin());
```
Do we need to do this copy?  That is, instead of normalizing into a temporary buffer and then copying into the `bufferptr_`, can we just normalize directly into `bufferptr_`?  That probably requires changes to `normalize` and `createNewNormalizedBuffer`, but it should be a lot more performant.

The same comment goes for the code below.

ahcorde replied:
fixed https://github.com/ros2/rviz/pull/726/commits/401b05200ec9f260349be1a04b947c0abc5db40d

ahcorde committed:
Avoid make copies

Signed-off-by: ahcorde <ahcorde@gmail.com>

ahcorde committed:
make linters happy

Signed-off-by: ahcorde <ahcorde@gmail.com>

ahcorde commented:
friendly ping @clalancette

clalancette commented on rviz_default_plugins/src/rviz_default_plugins/displays/image/ros_image_texture.cpp:
```diff
@@ -418,20 +422,34 @@ ImageData ROSImageTexture::setFormatAndNormalizeDataIfNecessary(ImageData image_
   {
     image_data.size_ /= sizeof(uint16_t);
-    std::vector<uint8_t> buffer = normalize<uint16_t>(
+    if (!bufferptr_) {
+      bufferptr_ = std::make_shared<std::vector<uint8_t>>(image_data.size_);
+    } else if (static_cast<size_t>(bufferptr_->size()) != image_data.size_) {
+      bufferptr_->resize(image_data.size_, 0);
+    }
```
A few different things about this code:

1. I feel like we can simplify this idiom by allocating at least one byte to `bufferptr_` in the constructor.  Then all of the places we do this, we can change this to:
```
if (bufferptr_->size() != image_data.size_) {
  bufferptr_->resize(image_data.size_, 0);
}
```
2. I don't think we need the static_cast to size_t; both `bufferptr_->size()` and `image_data.size_` are `size_t`.
3. Technically, we only need to resize the buffer if it is *smaller* than what we currently need.  That could save some reallocation, at the expense of some wasted memory.  I don't feel strongly about this point, though.
4. I can't quite put my finger on it, but the resize logic looks incorrect.  In particular, just prior to this code, we set `image_data.size_` to the number of uint16_t elements in the incoming piece of data.  But then we allocate that number of *uint8_t* bytes, which seems like we wouldn't have enough data in the bufferptr.  For example, assume that we have 10 uint16_t elements.  In that case, we'll allocate 10 bytes to the bufferptr, not 20.  It feels like we can overrun this when we are normalizing, but I'm not sure.

ahcorde replied:
I think these two lines are calculating the uint_8 size of the buffer.

```cpp
image_data.size_ /= sizeof(uint16_t);
...
image_data.size_ /= sizeof(float);
```

gbalke replied:
I can't see the reason to divide here. From looking at the origin of the data, we'd just want to allocate `image_data.size_`? I believe it's already the number of bytes not the number of elements? 

https://github.com/ros2/rviz/blob/018de8f730f3d9857395b7345acb5b8c55f38656/rviz_default_plugins/src/rviz_default_plugins/displays/image/ros_image_texture.cpp#L235

https://github.com/ros2/common_interfaces/blob/a3a0dde2ba184b01cdc59a3003728906de3240a9/sensor_msgs/msg/Image.msg#L26

ahcorde replied:
We need to do this because we are creating an image to show in RVIZ which is a `uint8_t`. We need to know the number of elements for each one of the formats.

Take a look to the `normalize()` function.

clalancette commented on rviz_default_plugins/src/rviz_default_plugins/displays/image/ros_image_texture.cpp:
```diff
@@ -173,16 +175,18 @@ double ROSImageTexture::computeMedianOfBuffer(const std::deque<double> & buffer)
+    buffer->resize(image_data_size, 0);
+  }
+
+  uint8_t * output_ptr = &((*buffer.get())[0]);
```
Is it possible to simplify this to:

```
uint8_t * output_ptr = &buffer[0];
```

ahcorde replied:
It's a shared_ptr, not really sure that I can do that

clalancette replied:
Ah, good point.  OK, never mind this comment then.

clalancette reviewed (commented):
I've left some comments inline.  One other thing that I noticed in this piece of code is that it isn't clear to me what `mutex_` is protecting.  Is it just the `image_` and `new_image_`?  Is it all of the variables in the class?  What can run on different threads here?  I don't know the answer to all of that, but I think it is worth investigating to see if we need to fix up locking here.

ahcorde committed:
Added feedback

Signed-off-by: ahcorde <ahcorde@gmail.com>

ros-discourse commented:
This pull request has been mentioned on **ROS Discourse**. There might be relevant details there:

https://discourse.ros.org/t/ros-2-tsc-meeting-minutes-2021-9-16/22372/1

ahcorde commented:
friendly ping @clalancette

clalancette committed:
Heavily revamp ros_image_texture.cpp

The main goal here is to stop using a pointer
from a freed std::vector.  Instead, we allocate
memory on the heap and delete it when this piece
of code no longer needs it (Ogre copies it internally).

Along the way, we do a huge cleanup of this code to
make it easier to understand and fix some obvious
problems with it:

1. Improve the locking.
2. Make some methods const.
3. Move implementations in the C++ file.
4. Remove unnecessary functions.
5. Rename a few things to make their function clearer.

Signed-off-by: Chris Lalancette <clalancette@gmail.com>

clalancette commented:
@ahcorde I know this one has been sitting a long time.  Sorry about that.

I ended up taking some time to look at this in more detail, and I ended up with a much larger rewrite here.  However, I believe that this rewrite takes things in the correct direction.

In particular, what we do here is to only store copied data into the ImageData as needed, and only for as long as needed.  That is enabled by the fact that Ogre::Image::load_raw_data() actually copies the data into itself.  Thus, we only need to keep that data around long enough for the function to be called, and we can then free it.  Further, we recognize that we only *sometimes* need to do a copy (really only when doing a conversion); otherwise, we can point to the data already stored in the message.

Beyond that, I did a bunch of other cleanup to make this a lot easier to understand.  I now think that this is in good shape, but I'd like you to take a look and let me know what you think about it.

ahcorde commented:
* Linux [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux&build=19810)](http://ci.ros2.org/job/ci_linux/19810/)
* Linux-aarch64 [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux-aarch64&build=14313)](http://ci.ros2.org/job/ci_linux-aarch64/14313/)
* Windows [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_windows&build=20534)](http://ci.ros2.org/job/ci_windows/20534/)

clalancette committed:
Fix an incorrect cast.

Signed-off-by: Chris Lalancette <clalancette@gmail.com>

clalancette commented:
There was a warning on Windows (which showed a real bug).  That is now fixed.  Here is new CI:

* Linux [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux&build=19813)](http://ci.ros2.org/job/ci_linux/19813/)
* Linux-aarch64 [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux-aarch64&build=14315)](http://ci.ros2.org/job/ci_linux-aarch64/14315/)
* Windows [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_windows&build=20536)](http://ci.ros2.org/job/ci_windows/20536/)
